client/client.go

```diff
@@ -90,6 +112,89 @@ func FetchCheckpoint(ctx context.Context, f CheckpointFetcherFunc, v note.Verifi
 	return cp, cpRaw, n, nil
 }
 
+// CheckConsistency is a wapper function which simplifies verifying consistency between two or more checkpoints.
+func CheckConsistency(ctx context.Context, f TileFetcherFunc, cp []log.Checkpoint) error {
+	if l := len(cp); l < 2 {
+		return fmt.Errorf("passed %d checkpoints, need at least 2", l)
+	}
+	sort.Slice(cp, func(i, j int) bool {
+		return cp[i].Size < cp[j].Size
+	})
+	pb, err := NewProofBuilder(ctx, cp[len(cp)-1], f)
+	if err != nil {
+		return fmt.Errorf("failed to create proofbuilder: %v", err)
+	}
+
+	// Go through list of checkpoints pairwise, checking consistency.
+	a, b := cp[0], cp[1]
+	for i := 0; i < len(cp)-1; i, a, b = i+1, cp[i], cp[i+1] {
+		if a.Size == b.Size {
+			if bytes.Equal(a.Hash, b.Hash) {
+				continue
+			}
+			return fmt.Errorf("two checkpoints with same size (%d) but different hashes (%x vs %x)", a.Size, a.Hash, b.Hash)
+		}
+		if a.Size > 0 {
+			cp, err := pb.ConsistencyProof(ctx, a.Size, b.Size)
+			if err != nil {
+				return fmt.Errorf("failed to fetch consistency between sizes %d, %d: %v", a.Size, b.Size, err)
+			}
+			if err := proof.VerifyConsistency(hasher, a.Size, b.Size, cp, a.Hash, b.Hash); err != nil {
+				return fmt.Errorf("invalid consistency proof between sizes %d, %d: %v", a.Size, b.Size, err)
+			}
+		}
+	}
+	return nil
+}
+
+// FetchRangeNodes returns the set of nodes representing the compact range covering
+// a log of size s.
+func FetchRangeNodes(ctx context.Context, s uint64, f TileFetcherFunc) ([][]byte, error) {
+	nc := newNodeCache(f, s)
+	nIDs := make([]compact.NodeID, 0, compact.RangeSize(0, s))
+	nIDs = compact.RangeNodes(0, s, nIDs)
+	hashes := make([][]byte, 0, len(nIDs))
+	for _, n := range nIDs {
+		h, err := nc.GetNode(ctx, n)
```

Comment:
Just a note that this has the potential to be a bit slow - the range nodes are likely to be in different tiles (given a non-trivial tree) and we're going to pull each required tile strictly sequentially.

Reply:
This is existing code. Maybe take a look at this diff commit by commit. This was moved in the first commit, which was just reordering stuff in the file.

Optimizing all the code in this file is out of scope for this PR.